pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -407,34 +408,43 @@ func (pl *Plugin) filterWithReservations(ctx context.Context, cycleState *framew
 			allocated = quotav1.Mask(allocated, rInfo.ResourceNames)
 			rRemained := quotav1.SubtractWithNonNegativeResult(rInfo.Allocatable, allocated)
 			requests := quotav1.Mask(state.podRequests, rInfo.ResourceNames)
-			fits, _ := quotav1.LessThanOrEqual(requests, rRemained)
+			fits, insufficientResourcesByReservation := quotav1.LessThanOrEqual(requests, rRemained)
 			if fits && nodeFits {
 				hasSatisfiedReservation = true
 				break
 			}
+			insufficientResources.Insert(quotav1.ToSet(insufficientResourcesByNode).List()...)
+			insufficientResources.Insert((quotav1.ToSet(insufficientResourcesByReservation)).List()...)
 		}
 	}
 	// The Pod requirement must be allocated from Reservation, but currently no Reservation meets the requirement
 	if !hasSatisfiedReservation && requiredFromReservation {
-		return framework.NewStatus(framework.Unschedulable, ErrReasonNoReservationsMeetRequirements)
+		// We will keep all failure reasons.
+		failureReasons := make([]string, 0, len(insufficientResources))
+		for insufficientResource := range insufficientResources {
+			failureReasons = append(failureReasons, fmt.Sprintf("Insufficient reservation %s", insufficientResource))
+		}
+		return framework.NewStatus(framework.Unschedulable, failureReasons...)
 	}
 	return nil
 }
 
 var dummyResource = framework.NewResource(nil)
 
 // fitsNode checks if node have enough resources to host the pod.
-func fitsNode(podRequest *framework.Resource, nodeInfo *framework.NodeInfo, nodeRState *nodeReservationState, rInfo *frameworkext.ReservationInfo, preemptible *framework.Resource) bool {
+func fitsNode(podRequest *framework.Resource, nodeInfo *framework.NodeInfo, nodeRState *nodeReservationState, rInfo *frameworkext.ReservationInfo, preemptible *framework.Resource) []corev1.ResourceName {
+	insufficientResources := make([]corev1.ResourceName, 0, 4)
```

Comment:
It is not recommended to initialize slices here as it is not needed in most cases.